bin/propolis-server/src/lib/migrate/compat.rs

```diff
@@ -0,0 +1,791 @@
+
+    #[error("collection {0} incompatible")]
+    Collection(String, #[source] CollectionIncompatibility),
+
+    #[cfg(feature = "falcon")]
+    #[error("can't migrate instances containing softnpu devices")]
+    SoftNpu,
+}
+
+#[derive(Debug, Error)]
+pub enum BoardIncompatibility {
+    #[error("boards have different CPU counts (self: {this}, other: {other})")]
+    CpuCount { this: u8, other: u8 },
+
+    #[error(
+        "boards have different memory sizes (self: {this}, other: {other})"
+    )]
+    MemorySize { this: u64, other: u64 },
+
+    #[error(
+        "chipsets have different PCIe settings (self: {this}, other: {other})"
+    )]
+    PcieEnabled { this: bool, other: bool },
+}
+
+#[derive(Debug, Error)]
+pub enum DiskIncompatibility {
+    #[error(
+        "disks have different device interfaces (self: {this}, other: {other})"
+    )]
+    Interface { this: &'static str, other: &'static str },
+
+    #[error("disks have different PCI paths (self: {this}, other: {other})")]
+    PciPath { this: PciPath, other: PciPath },
+
+    #[error(
+        "disks have different backend names (self: {this}, other: {other})"
+    )]
+    BackendName { this: String, other: String },
+
+    #[error(
+        "disks have different backend kinds (self: {this}, other: {other})"
+    )]
+    BackendKind { this: &'static str, other: &'static str },
+
+    #[error(
+        "disks have different read-only settings (self: {this}, other: {other})"
+    )]
+    ReadOnly { this: bool, other: bool },
+}
+
+#[derive(Debug, Error)]
+pub enum NicIncompatibility {
+    #[error("NICs have different PCI paths (self: {this}, other: {other})")]
+    PciPath { this: PciPath, other: PciPath },
+
+    #[error(
+        "NICs have different backend names (self: {this}, other: {other})"
+    )]
+    BackendName { this: String, other: String },
+}
+
+#[derive(Debug, Error)]
+pub enum SerialPortIncompatibility {
+    #[error("ports have different numbers (self: {this:?}, other: {other:?})")]
+    Number { this: SerialPortNumber, other: SerialPortNumber },
+
+    #[error("ports have different devices (self: {this}, other: {other})")]
+    Device { this: SerialPortDevice, other: SerialPortDevice },
+}
+
+#[derive(Debug, Error)]
+pub enum BridgeIncompatibility {
+    #[error("bridges have different PCI paths (self: {this}, other: {other})")]
+    PciPath { this: PciPath, other: PciPath },
+
+    #[error("bridges have different downstream buses (self: {this}, other: {other})")]
+    DownstreamBus { this: u8, other: u8 },
+}
+
+#[derive(Debug, Error)]
+pub enum PvpanicIncompatibility {
+    #[error("pvpanic presence differs (self: {this}, other: {other})")]
+    Presence { this: bool, other: bool },
+
+    #[error(
+        "pvpanic devices have different names (self: {this}, other: {other})"
+    )]
+    Name { this: String, other: String },
+
+    #[error(
+        "pvpanic devices have different ISA settings (self: {this}, other: {other})"
```

Comment:
super pedantic, feel free to ignore: the QEMU docs always typeset it as "PVPANIC" (all caps), so the error messages I had added previously did so as well; perhaps we should here? i don't actually care about this though. :)

Reply:
I'm gonna leave this one as-is--I know it's the PANIC DEVICE but IMO there's not really a need to shout ;) (If the name changes to the `Panic!` device I might change my opinion though.)

Reply:
yeah, i'm fine with that, i really don't care about typesetting it the same way as QEMU